File: sdks/python/src/stepflow_sdk/value.py
# Licensed to the Apache Software Foundation (ASF) under one or more contributor
# license agreements.  See the NOTICE file distributed with this work for
# additional information regarding copyright ownership.  The ASF licenses this
# file to you under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

"""Value API for creating workflow values and references."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .generated_protocol import (
    EscapedLiteral,
    Reference,
    SkipAction,
    ValueTemplate,
    WorkflowRef,
    WorkflowReference,
)
from .generated_protocol import (
    StepReference as GeneratedStepReference,
)


class JsonPath:
    """Base class for handling JSON Path syntax across all reference types."""

    def __init__(self):
        """Initialize with an optional initial path."""
        self.fragments = ["$"]

    def push_field(self, name: str) -> JsonPath:
        """Add a field access to the path (e.g., .fieldName). Mutates this instance."""
        self.fragments.append(f".{name}")
        return self

    def push_index(self, key: str | int) -> JsonPath:
        """
        Add an index access to the path (e.g., [0] or ["key"]).

        Mutates this instance.
        """
        if isinstance(key, int):
            self.fragments.append(f"[{key}]")
        else:
            self.fragments.append(f'["{key}"]')
        return self

    def with_field(self, name: str) -> JsonPath:
        path = self.copy()
        path.push_field(name)
        return path

    def with_index(self, key: str | int) -> JsonPath:
        path = self.copy()
        path.push_index(key)
        return path

    def copy(self) -> JsonPath:
        """Create a copy of this JsonPath instance."""
        new_path = JsonPath()
        new_path.fragments = self.fragments.copy()
        return new_path

    def __str__(self) -> str:
        """Return the path string, defaulting to '$' for root."""
        return "".join(self.fragments)

    def __repr__(self) -> str:
        return f"JsonPath({self.fragments!r})"


@dataclass
class StepReference:
    """A reference to a step's output or a field within it."""

    step_id: str
    path: JsonPath = field(default_factory=JsonPath)
    on_skip: SkipAction | None = None

    def __getitem__(self, key: str | int) -> StepReference:
        """Create a nested reference using index access."""
        new_path = self.path.with_index(key)
        return StepReference(self.step_id, new_path, self.on_skip)

    def __getattr__(self, name: str) -> StepReference:
        """Create a nested reference to a field."""
        new_path = self.path.with_field(name)
        return StepReference(self.step_id, new_path, self.on_skip)

    def with_on_skip(self, on_skip: SkipAction) -> StepReference:
        """Create a copy of this reference with the specified onSkip action."""
        return StepReference(self.step_id, self.path, on_skip)


@dataclass
class WorkflowInput:
    """Reference to workflow input."""

    path: JsonPath = field(default_factory=JsonPath)
    on_skip: SkipAction | None = None

    def __getitem__(self, key: str | int) -> WorkflowInput:
        """Create a reference to a specific path in the workflow input."""
        new_path = self.path.with_index(key)
        return WorkflowInput(new_path, self.on_skip)

    def __getattr__(self, name: str) -> WorkflowInput:
        """Create a reference to a field in the workflow input."""
        new_path = self.path.with_field(name)
        return WorkflowInput(new_path, self.on_skip)

    def with_on_skip(self, on_skip: SkipAction) -> WorkflowInput:
        """Create a copy of this reference with the specified onSkip action."""
        return WorkflowInput(self.path, on_skip)


class Value:
    """A value that can be used in workflow definitions.

    This class provides a unified interface for creating values that can be:
    - Literal values (using Value.literal() or Value())
    - References to steps (using Value.step())
    - References to workflow input (using Value.input())
    - Expressions using $literal and $from syntax
    """

    def __init__(self, value: Valuable):
        """Create a Value from a Valuable.

        Args:
            value: The value to wrap. Can be:
                - A literal value (str, int, float, bool, None, dict, list)
                - A StepReference or WorkflowInput
                - An EscapedLiteral
                - Another Value (will be unwrapped)
        """
        self._value = self._unwrap_value(value)

    def _unwrap_value(self, value: Valuable) -> Any:
        """Unwrap a Value if it's wrapped, otherwise return as-is."""
        if isinstance(value, Value):
            return value._value
        return value

    @staticmethod
    def literal(value: Any) -> Value:
        """Create a literal value that won't be expanded as a reference.

        This is equivalent to using $literal in the workflow definition.
        """
        return Value(EscapedLiteral(field_literal=value))

    @staticmethod
    def step(
        step_id: str, path: str | None = None, on_skip: SkipAction | None = None
    ) -> Value:
        """Create a reference to a step's output.

        This is equivalent to using $from with a step reference.
        """
        json_path = JsonPath()
        if path is not None and path != "$":
            json_path.fragments = [path]
        return Value(StepReference(step_id, json_path, on_skip))

    @staticmethod
    def input(path: str | None = None, on_skip: SkipAction | None = None) -> Value:
        """Create a reference to workflow input.

        This is equivalent to using $from with a workflow input reference.
        """
        json_path = JsonPath()
        if path is not None and path != "$":
            json_path.fragments = [path]
        return Value(WorkflowInput(json_path, on_skip))

    def to_value_template(self) -> ValueTemplate:
        """Convert this Value to a ValueTemplate for use in flow definitions."""
        return Value._convert_to_value_template(self._value)

    @staticmethod
    def _convert_to_value_template(data: Any) -> ValueTemplate | None:
        """Convert arbitrary data to ValueTemplate."""
        if data is None:
            return None

        if isinstance(data, Value):
            return Value._convert_to_value_template(data._value)

        if isinstance(data, StepReference | WorkflowInput):
            return Value._convert_reference_to_expr(data)

        if isinstance(data, EscapedLiteral):
            return data

        if isinstance(data, dict):
            converted = {}
            for key, value in data.items():
                converted[key] = Value._convert_to_value_template(value)
            return converted

        if isinstance(data, list):
            return [Value._convert_to_value_template(item) for item in data]

        # For primitive types (str, int, float, bool), return as-is
        return data

    @staticmethod
    def _convert_reference_to_expr(
        ref: StepReference | WorkflowInput,
    ) -> Reference:
        """Convert a reference to a Reference."""
        base_ref: WorkflowReference | GeneratedStepReference
        if isinstance(ref, StepReference):
            base_ref = GeneratedStepReference(step=ref.step_id)
            path_str = str(ref.path) if str(ref.path) != "$" else None
            return Reference(field_from=base_ref, path=path_str, onSkip=ref.on_skip)
        elif isinstance(ref, WorkflowInput):
            base_ref = WorkflowReference(workflow=WorkflowRef.input)
            path_str = str(ref.path) if str(ref.path) != "$" else None
            return Reference(field_from=base_ref, path=path_str, onSkip=ref.on_skip)
        else:
            raise ValueError(f"Unknown reference type: {type(ref)}")

    def __getitem__(self, key: str | int) -> Value:
        """
        Create a nested reference to the given key.

        Raises an error if this is not a reference.
        """
        if isinstance(self._value, StepReference | WorkflowInput):
            return Value(self._value[key])
        else:
            raise TypeError(f"Cannot index into {type(self._value).__name__}")

    def __getattr__(self, name: str) -> Value:
        """
        Create a nested reference to the given name.

        Raises an error if this is not a reference.
        """
        if isinstance(self._value, StepReference | WorkflowInput):
            return Value(getattr(self._value, name))
        else:
            raise AttributeError(
                f"'{type(self._value).__name__}' object has no attribute '{name}'"
            )

    def with_on_skip(self, on_skip: SkipAction) -> Value:
        """Create a copy of this Value with the specified onSkip action (only for
        references)."""
        if isinstance(self._value, StepReference | WorkflowInput):
            return Value(self._value.with_on_skip(on_skip))
        else:
            raise TypeError(f"Cannot set onSkip on {type(self._value).__name__}")


class WorkflowInputValue(Value):
    """Special Value class for workflow input that supports attribute access."""

    def __init__(self):
        super().__init__(WorkflowInput())

    def __getitem__(self, key: str | int) -> Value:
        """Create a nested reference to workflow input."""
        return Value(self._value[key])

    def __getattr__(self, name: str) -> Value:
        """Create a nested reference to workflow input field."""
        return Value(getattr(self._value, name))


# Type alias for things that can be converted to JSON values
Valuable = (
    Value
    | StepReference
    | WorkflowInput
    | EscapedLiteral
    | str
    | int
    | float
    | bool
    | None
    | dict[str, "Valuable"]
    | list["Valuable"]
)

File: sdks/typescript/src/value.ts
/**
 * Value API for creating workflow values and references.
 * This provides a TypeScript implementation of the Python SDK's Value system.
 */

import { ValueTemplate, ReferenceExpr, EscapedLiteralExpr, SkipAction } from './protocol';

export type Valuable = 
  | string
  | number
  | boolean
  | null
  | undefined
  | Record<string, any>
  | Array<any>
  | StepReference
  | WorkflowInput
  | EscapedLiteral
  | Value;

export interface EscapedLiteral {
  field_literal: any;
}

// Re-export protocol types for convenience
export { SkipAction } from './protocol';

/**
 * Base class for handling JSON Path syntax across all reference types.
 */
export class JsonPath {
  private fragments: string[] = ['$'];

  constructor(fragments?: string[]) {
    if (fragments) {
      this.fragments = [...fragments];
    }
  }

  /**
   * Add a field access to the path (e.g., .fieldName). Mutates this instance.
   */
  pushField(name: string): JsonPath {
    this.fragments.push(`.${name}`);
    return this;
  }

  /**
   * Add an index access to the path (e.g., [0] or ["key"]). Mutates this instance.
   */
  pushIndex(key: string | number): JsonPath {
    if (typeof key === 'number') {
      this.fragments.push(`[${key}]`);
    } else {
      this.fragments.push(`["${key}"]`);
    }
    return this;
  }

  /**
   * Create a new path with a field access.
   */
  withField(name: string): JsonPath {
    return this.copy().pushField(name);
  }

  /**
   * Create a new path with an index access.
   */
  withIndex(key: string | number): JsonPath {
    return this.copy().pushIndex(key);
  }

  /**
   * Create a copy of this JsonPath instance.
   */
  copy(): JsonPath {
    return new JsonPath([...this.fragments]);
  }

  /**
   * Return the path string, defaulting to '$' for root.
   */
  toString(): string {
    return this.fragments.join('');
  }

  /**
   * Get the fragments array.
   */
  getFragments(): string[] {
    return [...this.fragments];
  }

  /**
   * Set the entire path from a string.
   */
  setPath(path: string): JsonPath {
    this.fragments = [path];
    return this;
  }
}

/**
 * A reference to a step's output or a field within it.
 */
export class StepReference {
  constructor(
    public readonly stepId: string,
    public readonly path: JsonPath = new JsonPath(),
    public readonly onSkip?: SkipAction
  ) {}

  /**
   * Create a nested reference using index access.
   */
  [Symbol.toPrimitive](hint: string): any {
    if (hint === 'string') {
      return `StepReference(${this.stepId}, ${this.path.toString()})`;
    }
    return this;
  }

  /**
   * Create a nested reference using bracket notation.
   */
  get(key: string | number): StepReference {
    const newPath = this.path.withIndex(key);
    return new StepReference(this.stepId, newPath, this.onSkip);
  }

  /**
   * Create a nested reference to a field using a proxy for property access.
   */
  private static createProxy(ref: StepReference): StepReference {
    return new Proxy(ref, {
      get(target, prop) {
        // Preserve existing methods
        if (prop in target || typeof prop === 'symbol') {
          return (target as any)[prop];
        }
        
        if (typeof prop === 'string' && !prop.startsWith('_')) {
          const newPath = target.path.withField(prop);
          return StepReference.createProxy(new StepReference(target.stepId, newPath, target.onSkip));
        }
        return (target as any)[prop];
      }
    });
  }

  /**
   * Create a new StepReference with proxy support for field access.
   */
  static create(stepId: string, path?: JsonPath, onSkip?: SkipAction): StepReference {
    const ref = new StepReference(stepId, path, onSkip);
    return StepReference.createProxy(ref);
  }

  /**
   * Create a copy of this reference with the specified onSkip action.
   */
  withOnSkip(onSkip: SkipAction): StepReference {
    return new StepReference(this.stepId, this.path, onSkip);
  }

  /**
   * Convert to ReferenceExpr object for protocol.
   */
  toReferenceExpr(): ReferenceExpr {
    const expr: ReferenceExpr = {
      $from: { step: this.stepId }
    };
    
    const pathStr = this.path.toString();
    if (pathStr !== '$') {
      expr.path = pathStr;
    }
    
    if (this.onSkip) {
      expr.onSkip = this.onSkip;
    }
    
    return expr;
  }

  /**
   * @deprecated Use toReferenceExpr() instead
   */
  toReference(): any {
    return {
      step: {
        step_id: this.stepId,
        path: this.path.toString(),
        on_skip: this.onSkip
      }
    };
  }
}

/**
 * Reference to workflow input.
 */
export class WorkflowInput {
  constructor(
    public readonly path: JsonPath = new JsonPath(),
    public readonly onSkip?: SkipAction
  ) {}

  /**
   * Create a reference to a specific path in the workflow input using bracket notation.
   */
  get(key: string | number): WorkflowInput {
    const newPath = this.path.withIndex(key);
    return new WorkflowInput(newPath, this.onSkip);
  }

  /**
   * Create a nested reference to a field using a proxy for property access.
   */
  private static createProxy(input: WorkflowInput): WorkflowInput {
    return new Proxy(input, {
      get(target, prop) {
        // Preserve existing methods
        if (prop in target || typeof prop === 'symbol') {
          return (target as any)[prop];
        }
        
        if (typeof prop === 'string' && !prop.startsWith('_')) {
          const newPath = target.path.withField(prop);
          return WorkflowInput.createProxy(new WorkflowInput(newPath, target.onSkip));
        }
        return (target as any)[prop];
      }
    });
  }

  /**
   * Create a new WorkflowInput with proxy support for field access.
   */
  static create(path?: JsonPath, onSkip?: SkipAction): WorkflowInput {
    const input = new WorkflowInput(path, onSkip);
    return WorkflowInput.createProxy(input);
  }

  /**
   * Create a copy of this reference with the specified onSkip action.
   */
  withOnSkip(onSkip: SkipAction): WorkflowInput {
    return new WorkflowInput(this.path, onSkip);
  }

  /**
   * Convert to ReferenceExpr object for protocol.
   */
  toReferenceExpr(): ReferenceExpr {
    const expr: ReferenceExpr = {
      $from: { workflow: "input" }
    };
    
    const pathStr = this.path.toString();
    if (pathStr !== '$') {
      expr.path = pathStr;
    }
    
    if (this.onSkip) {
      expr.onSkip = this.onSkip;
    }
    
    return expr;
  }

  /**
   * @deprecated Use toReferenceExpr() instead
   */
  toReference(): any {
    return {
      input: {
        path: this.path.toString(),
        on_skip: this.onSkip
      }
    };
  }
}

/**
 * A value that can be used in workflow definitions.
 * 
 * This class provides a unified interface for creating values that can be:
 * - Literal values (using Value.literal() or new Value())
 * - References to steps (using Value.step())
 * - References to workflow input (using Value.input())
 */
export class Value {
  private _value: any;

  constructor(value: Valuable) {
    this._value = this.unwrapValue(value);
  }

  private unwrapValue(value: Valuable): any {
    if (value instanceof Value) {
      return value._value;
    }
    return value;
  }

  /**
   * Create a literal value that won't be expanded as a reference.
   * This is equivalent to using $literal in the workflow definition.
   */
  static literal(value: any): Value {
    return new Value({ field_literal: value } as EscapedLiteral);
  }

  /**
   * Create a reference to a step's output.
   * This is equivalent to using $from with a step reference.
   */
  static step(stepId: string, path?: string, onSkip?: SkipAction): Value {
    const jsonPath = new JsonPath();
    if (path && path !== '$') {
      jsonPath.setPath(path);
    }
    return new Value(new StepReference(stepId, jsonPath, onSkip));
  }

  /**
   * Create a reference to workflow input.
   * This is equivalent to using $from with a workflow input reference.
   */
  static input(path?: string, onSkip?: SkipAction): Value {
    const jsonPath = new JsonPath();
    if (path && path !== '$') {
      jsonPath.setPath(path);
    }
    return new Value(WorkflowInput.create(jsonPath, onSkip));
  }

  /**
   * Convert this Value to a ValueTemplate for use in flow definitions.
   */
  toValueTemplate(): ValueTemplate {
    return Value.convertToValueTemplate(this._value);
  }

  /**
   * Convert arbitrary data to ValueTemplate.
   */
  static convertToValueTemplate(data: any): ValueTemplate {
    if (data === null || data === undefined) {
      return data;
    }

    if (data instanceof Value) {
      return Value.convertToValueTemplate(data._value);
    }

    if (data instanceof StepReference) {
      return data.toReferenceExpr();
    }

    if (data instanceof WorkflowInput) {
      return data.toReferenceExpr();
    }

    if (data && typeof data === 'object' && 'field_literal' in data) {
      return { $literal: (data as EscapedLiteral).field_literal };
    }

    if (Array.isArray(data)) {
      return data.map(item => Value.convertToValueTemplate(item));
    }

    if (data && typeof data === 'object') {
      const result: Record<string, any> = {};
      for (const [key, value] of Object.entries(data)) {
        result[key] = Value.convertToValueTemplate(value);
      }
      return result;
    }

    return data;
  }

  /**
   * Get the underlying value.
   */
  getValue(): any {
    return this._value;
  }

  /**
   * Create a nested reference using bracket notation (when Value wraps a reference).
   */
  get(key: string | number): Value {
    if (this._value instanceof StepReference) {
      return new Value(this._value.get(key));
    }
    if (this._value instanceof WorkflowInput) {
      return new Value(this._value.get(key));
    }
    throw new Error('get() can only be called on Value instances that wrap StepReference or WorkflowInput');
  }
}